### PR TITLE
DP-InstaHide Trainer Implementation

### LINKED ISSUE #1827: Implementation of DP-InstaHide and Data Augmentation Defenses
**Is your feature request related to a problem? Please describe.**
Mixup, CutOut, CutMix, and Maxup are relatively simple data augmentation defenses that have been shown to provide robustness against poisoning attacks. ART currently does not have any data augmentation defenses so this implementation will provide more features and open the doors for additional data augmentation defenses. DP-InstaHide is a relatively simple training method that utilizes data augmentation defenses and has privacy guarantees. This implementation will be a good utilization of the data augmentation defenses and provide another useful tool to ART.

**Describe the solution you'd like**
Implementation of data augmentation defenses would involve creation of the `data_augmentation` submodule under `art.defenses.preprocessor`. Here, an abstract class for data augmentation would be created that follows the same structure as the preprocessor defenses. This abstract class would then be extended to create defenses for `Mixup`, `CutOut`, `CutMix`, and `Maxup` for the four main data augmentation defenses used with DP-InstaHide.

Implementation of DP-InstaHide would involve creating a new trainer under `art.defenses.trainer` which would run similarly to the adversarial trainer. This would take parameters to specify the type of data augmentation defense along with the hyperparameters.

**Describe alternatives you've considered**
Originally, the idea was to have `data_augmentation` be its own submodule under `art.defenses` rather than a type of preprocessor. However, after some thought and discussion, the data augmentation defenses should be a submodule under `art.defenses.preprocessor`.

**Additional context**
It might be worth splitting this into two separate issues for organization purposes. One for the implementation of data augmentation defenses and another for the implementation of DP-InstaHide.

### LINKED ISSUE #1959: TensorFlowV2Classifier and KerasClassifier do not validate labels when computing loss
**Describe the bug**
In the current implementations for the `TensorFlowV2Classifier` and `KerasClassifier`, when calling the `compute_loss(x, y)` method, the labels are not validated and transformed. This causes an error when the `argmax` function is called and integer class labels are passed.

The `PyTorchClassifier` class performs this validation so the error is avoided. The fix is relatively straightforward, just call the `check_and_transform_label_format` function in the `TensorFlowV2Classifier` and `KerasClassifier` classes just like is done with the PyTorch variation.

**To Reproduce**
Instantiate a `TensorFlowV2Classifer` or `KerasClassifier` instance using the `SparseCategoricalCrossentropy` loss. Call the `compute_loss(x, y)` function and use labels for `y` (e.g., shape `(n,)`) and there will be an error due the argmax function call.

**Expected behavior**
The labels should be validated and transformed accordingly.

**System information (please complete the following information):**
 - OS
 - Python version
 - ART version or commit number
 - TensorFlow / Keras / PyTorch / MXNet version

### LINKED ISSUE #1966: Support Probability Distribution Labels in PyTorchClassifier
**Is your feature request related to a problem? Please describe.**
PyTorch's `nn.CrossEntropyLoss` function supports probability distribution labels. However, since the `PyTorchClassifier` always sets `_reduce_labels = True` when using this loss function, it is not possible to use probability distribution labels. The solution is to never reduce labels for this loss function as the result is the same when using vector labels or one-hot encoded labels.

**Describe the solution you'd like**
When calling `classifier.fit()` or `classifier.compute_loss()` using `nn.CrossEntropyLoss`, labels passed as a probability distribution should be kept as is and not reduced. No clear solution has been decided yet, instead ideas are below.

**Describe alternatives you've considered**
Create a new `_probability_labels` property in the `PyTorchClassifier` and have it set to True when using `nn.CrossEntropyLoss`. This will allow to check in the `reduce_labels()` function if the provided labels are one-hot encoded or probability distribution labels. However, this would introduce additional computation time to check the labels and may also break some of the tests.

Manually checking the labels beforehand to determine whether to reduce labels. But this would introduce extra complexity and everything should be left inside the `reduce_labels()` function.

Setting the `_reduce_labels` property to False in the `PyTorchClassifier` when using `nn.CrossEntropyLoss` so labels are always reduced is also an option since the results are the same. However, this breaks some of the unit tests for the PyTorch randomized smoothing implementation.

Use the `kwargs` in the `.fit()` and `.compute_loss()` functions to pass parameters to override the current `_reduce_labels` and `_int_labels` fields. This will preserve compatibility, but is not the most elegant solution.

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by f4str**
# Description

Implementation of the DP-InstaHide trainer. This trainer builds upon the previously implemented data augmentation defenses. Also includes a jupyter notebook demo showing how to use the trainer and its effectiveness on MNIST (a reference was added to the notebook README).

As a consequence, the following two bugs/issues were resolved
* Call the `check_and_transform_label_format` function when calling the `compute_loss` function in the `TensorFlowV2Classifier` and `KerasClassifier` classes. This fixes the argmax issue when passing integer labels. (Consolidated from PR #1960)
* Create a new property `_probability_labels` in the `PyTorchClassifier` and have it set to True when using `nn.CrossEntropyLoss`. This allows the `reduce_labels()` function to check if the labels are one-hot encoded or a probability distribution and handle it accordingly. This allows support for probability distribution labels. (Consolidated from PR #1967)

Fixes #1827 
Fixes #1959 
Fixes #1966

## Type of change

Please check all relevant options.

- [x] Improvement (non-breaking)
- [x] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] Test the DP-InstaHide trainer in PyTorch
- [x] Test the DP-InstaHide trainer in TensorFlow
- [x] Test the DP-InstaHide trainer in Keras

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by f4str** [comment]
The implementation and unit tests are complete and ready for review, but will also be adding a notebook demo.

**Comment by beat-buesser** [comment]
A review comment, please also add the new notebook to the notebook README file.

File: art/defences/trainer/dp_instahide_trainer.py [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+                x_aug = x_batch.copy()
+                y_aug = y_batch.copy()
+                for augmentation in self.augmentations:
+                    x_aug, y_aug = augmentation(x_aug, y_aug)
+
+                # apply additive noise
+                x_aug = self._generate_noise(x_aug)
+
+                # extract label reduction and set to no reduction if needed
+                reduce_labels = self._classifier._reduce_labels  # type: ignore # pylint: disable=W0212
```
Comment (beat-buesser) [inline art/defences/trainer/dp_instahide_trainer.py:158]: Why is there a type ignore?

File: art/defences/trainer/dp_instahide_trainer.py [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+                self._classifier.fit(x_aug, y_aug, nb_epochs=1, batch_size=x_aug.shape[0], verbose=0, **kwargs)
+
+                # get metrics
+                loss = self._classifier.compute_loss(x_aug, y_aug, reduction="mean")
+                output = np.argmax(self.predict(x_batch), axis=1)
+                acc = np.sum(output == np.argmax(y_batch, axis=1))
+                n = len(x_aug)
+
+                # restore original label reduction
+                self._classifier._reduce_labels = reduce_labels  # type: ignore # pylint: disable=W0212
```
Comment (beat-buesser) [inline art/defences/trainer/dp_instahide_trainer.py:259]: I think these lines should not be required anymore with https://github.com/Trusted-AI/adversarial-robustness-toolbox/pull/1960?

File: art/defences/trainer/dp_instahide_trainer.py [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+                for augmentation in self.augmentations:
+                    x_aug, y_aug = augmentation(x_aug, y_aug)
+
+                # apply additive noise
+                x_aug = self._generate_noise(x_aug)
+
+                # extract label reduction and set to no reduction if needed
+                reduce_labels = self._classifier._reduce_labels  # type: ignore # pylint: disable=W0212
+                if len(y_aug.shape) == 2:
+                    self._classifier._reduce_labels = False  # type: ignore # pylint: disable=W0212
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/defences/trainer/dp_instahide_trainer.py:247]: If yes, we need to change `.fit` and `.compute_loss`.

File: art/defences/trainer/dp_instahide_trainer.py (hunk 1 above)
Comment (f4str) [reply art/defences/trainer/dp_instahide_trainer.py:158]: Not all classifiers of type `CLASSIFIER_LOSS_GRADIENTS_TYPE` have the `_reduce_labels` property which causes a mypy error. This line is needed due to the `PyTorchClassifier` always reducing labels for `nn.CrossEntropyLoss`.

File: art/defences/trainer/dp_instahide_trainer.py (hunk 2 above)
Comment (f4str) [reply art/defences/trainer/dp_instahide_trainer.py:259]: Same as above

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Ho @f4str Thank you very much for contributing DP-InstaHide to ART!

**Comment by f4str** [comment]
This PR will be completed after PR #1960 and PR #1967 are merged

**Review by beat-buesser — APPROVED** [review]
Hi @f4str Thank you very much for contributing DP-InstaHide to ART! The review changes look good to me.
